Tools/swat/Doc/Tcl.man
'\" Copyright 1989 Regents of the University of California
'\" Permission to use, copy, modify, and distribute this
'\" documentation for any purpose and without fee is hereby
'\" granted, provided that this notice appears in all copies.
'\" The University of California makes no representations about
'\" the suitability of this material for any purpose.  It is
'\" provided "as is" without express or implied warranty.
'\" 
'\" $Header: /sprite/src/lib/tcl/RCS/Tcl.man,v 1.1 88/12/31 15:58:29 ouster Exp $ SPRITE (Berkeley)
'\"
'\".so \*(]ltmac.sprite
.nr #D 5n
.de DS   \" Real Display-Start macro. It actually works!
.sp .6v
.nf
.in +\\n(#Du
..
.de DE   \" Real Display-End macro.
.in
.fi
.sp .6v
..
.TH Tcl 3
'\"
'\" Those obnoxious twits at ISI don't let you put anything but their
'\" little banner at the bottom of the page, so redefine their footer
'\" macro to leave the ]W string alone.
'\"
.de }F
.ev 1
.ft 1
.ps 10
'sp .5i
.ds ]W "Sprite v1.0
.tl @\\*(]L@\\*(]W@%@
'bp
.ev
..
.SH NAME
Tcl \- overview of tool command language facilities

.SH INTRODUCTION
.PP
Tcl stands for ``tool command language'' and is pronounced ``tickle.''
It is actually two things:
a language and a library.
First, Tcl is a simple textual language,
intended primarily for issuing commands to interactive programs such
as text editors, debuggers, illustrators, and shells.  It has
a simple syntax and is also programmable, so
Tcl users can write command procedures to provide more powerful
commands than those in the built-in set.
.PP
Second, Tcl is a library package that can be embedded in application
programs.  The Tcl library consists of a parser for the Tcl
language, routines to implement the Tcl built-in commands, and
procedures that allow each application to extend Tcl with additional
commands specific to that application.  The application program
generates Tcl commands and passes them to the Tcl parser for
execution.  Commands may be generated
by reading characters from an input
source, or by associating command strings with elements of the
application's user interface, such as menu entries, buttons, or
keystrokes.
When the Tcl library receives commands it parses them
into component fields and executes built-in commands directly.
For commands implemented by the
application, Tcl calls back to the application to execute the
commands.  In many cases commands will invoke recursive invocations
of the Tcl interpreter by passing in additional strings to execute
(procedures, looping commands, and conditional commands all work
in this way).
.PP
An application program gains three advantages by using Tcl for
its command language.  First, Tcl provides a standard syntax:  once
users know Tcl, they will be able to issue commands easily
to any Tcl-based application.  Second, Tcl provides programmability.
All a Tcl application needs to do is to implement a few
application-specific low-level commands.  Tcl provides many utility
commands plus a general programming interface for building up
complex command procedures.  By using Tcl, applications need not
re-implement these features.  Third, Tcl will eventually provide
a mechanism for communicating between applications:  it will be
possible to send Tcl commands from one application to another.
The common Tcl language framework will make it easier for applications
to communicate with one another.  The communication features are not
implemented in the current version of Tcl.
.PP
This manual page focusses primarily on the Tcl language.  It describes
the language syntax and the built-in commands that will be available in
any application based on Tcl.  The individual library
procedures are described in more detail in separate manual pages, one
per procedure.

.SH "INTERPRETERS"
.PP
The central data structure in Tcl is an interpreter (C type
``Tcl_Interp'').  An interpreter consists of a set of command
bindings, a set of variable values, and a few other miscellaneous
pieces of state.  Each Tcl command is interpreted in the context
of a particular interpreter.
Some Tcl-based applications will maintain
multiple interpreters simultaneously, each associated with a
different widget or portion of the application.
Interpreters are relatively lightweight structures.  They can
be created and deleted quickly, so application programmers should feel free to
use multiple interpreters if that simplifies the application.
Eventually Tcl will provide a mechanism for sending Tcl commands
and results back and forth between interpreters, even if the
interpreters are managed by different processes.

.SH "DATA TYPES"
.PP
Tcl supports only one type of data:  strings.  All commands,
all arguments to commands, all command results, and all variable values
are strings.
Where commands require numeric arguments or return numeric results,
the arguments and results are passed as strings.
Many commands expect their string arguments to have certain formats,
but this interpretation is
up to the individual commands.  For example, arguments often contain
Tcl command strings, which may get executed as part of the commands.
The easiest way to understand the Tcl interpreter is to remember that
everything is just an operation on a string.  In many cases Tcl constructs
will look similar to more structured constructs from other languages.
However, the Tcl constructs
are not structured at all;  they are just strings of characters, and this
gives them a different behavior than the structures they may look like.
.PP
Although the exact interpretation of a Tcl string depends on who is
doing the interpretation, there are three common forms that strings
take:  commands, expressions, and lists.  The major sections below
discuss these three forms in more detail.

.SH "BASIC COMMAND SYNTAX"
.PP
The Tcl language has syntactic similarities to both the Unix shells
and Lisp.  However, the interpretation of commands is different
in Tcl than in either of those other two systems.
A Tcl command string consists of one or more commands separated
by newline characters.
Each command consists of a collection of fields separated by
white space (spaces or tabs).
The first field must be the name of a command, and the
additional fields, if any, are arguments that will be passed to
that command.  For example, the command
.DS
\fBvar a 22\fR
.DE
has three fields:  the first, \fBvar\fR, is the name of a Tcl command, and
the last two, \fBa\fR and \fB22\fR, will be passed as arguments to
the \fBvar\fR command.  The command name may refer either to a built-in
Tcl command, an application-specific command bound in with the library
procedure \fBTcl_CreateCommand\fR, or a command procedure defined with the
\fBproc\fR built-in command.
Arguments are passed literally as
text strings.  Individual commands may interpret those strings in any
fashion they wish.  The \fBvar\fR command, for example, will treat its
first argument as the name of a variable and its second argument as a
string value to assign to that variable.  For other commands arguments
may be interpreted as integers, lists, file names, or Tcl commands.

.SH "COMMENTS"
.PP
If the first non-blank character in a command is \fB#\fR, then everything
from the \fB#\fR up through the next newline character is treated as
a comment and ignored.

.SH "GROUPING ARGUMENTS WITH BRACES"
.PP
Normally each argument field ends at the next white space, but
curly braces (``{'' and ``}'') may be used to group arguments
in different ways.
If an argument field
begins with a left brace, then the argument isn't terminated by
white space;  it ends at the matching right brace.  Tcl will
strip off the outermost layer of braces before passing the
argument to the command.  For example, in the command
.DS
\fBvar a {xyz a {b c d}}\fR
.DE
the \fBvar\fR command will receive two arguments: \fBa\fR
and \fBxyz a {b c d}\fR.  The matching right brace need not be on
the same line as the left brace;  in this case the newline will be
included in the argument field along with any other characters up to the
matching right brace.  In many cases an argument field to one
command consists of a Tcl command string that will be executed later;
braces allow complex command structures to be built
up without confusion.  For example, the \fBeval\fR command takes one
argument, which is a command string;  \fBeval\fR invokes the Tcl
interpreter to execute the command string.  The command
.DS
\fBeval {
	var a 22
	var b 33
}\fR
.DE
will assign the value \fB22\fR to \fBa\fR and \fB33\fR to \fBb\fR.
.PP
Tcl braces act like quote
characters in most other langauges, in that they prevent any special
interpretation of the characters between the left brace and
the matching right brace.  Braces are better
than quotes because there are separate left and right braces, which
makes them easier to nest.
.PP
When an argument is in braces, then command, variable, and backslash
substitutions do not occur as described below;  all Tcl does is to
strip off the outer layer of braces and pass the contents to the
command.
.PP
Braces are only significant in a command field if the first character
of the field is a left brace.  Otherwise neither left nor right
braces in the field will be treated specially (except as part of
variable substitution;  see below). 

.SH "GROUPING COMMANDS WITH BRACKETS"
.PP
Normally, each command occupies one line (the command is terminated by
a newline character).  Thus, the string
.DS
\fBvar a 22
var b 33\fR
.DE
will be interpreted as two separate commands.  However, brackets may be
used to group commands in ways other than one-command-per-line.
If the first
character of a command is an open bracket then the command isn't
terminated by a newline character;  instead, it consists of all the characters
up to the matching close bracket.  Newline characters inside a
bracketed command are treated as white space (they will act as
argument separators for arguments that aren't enclosed in braces).
Thus, the string
.DS
\fB[var a
22] [var b 33]\fR
.DE
will produce the same effect as the version above.

.SH "COMMAND SUBSTITUTION WITH BRACKETS"
.PP
If an open bracket occurs in any of the fields of a command, then
command substitution occurs.  All of the text up to the matching
close bracket is treated as a Tcl command and executed immediately.
Then the result of that command is substituted for the bracketed
text.  For example, consider the command
.DS
\fBvar a [var b]\fR
.DE
When the \fBvar\fR command has only a single argument, it is the
name of a variable and \fBvar\fR returns the contents of that
variable.  In this case, if variable \fBb\fR has the value \fBfoo\fR,
then the command above is equivalent to the command
.DS
\fBvar a foo\fR
.DE
Brackets can be used in more complex ways.  For example, if the
variable \fBb\fR has the value \fBfoo\fR and the variable \fBc\fR
has the value \fBgorp\fR, then the command
.DS
\fBvar a xyz[var b].[var c]\fR
.DE
is equivalent to the command
.DS
\fBvar a xyzfoo.gorp\fR
.DE
If a field is enclosed in braces then the brackets and the characters
between them are not interpreted specially;  they are passed through
to the argument verbatim.

.SH "VARIABLE SUBSTITUTION WITH $"
.PP
The dollar sign (\fB$\fR) may be used as a special shorthand form
for substituting variables.  If \fB$\fR appears in an argument that
isn't enclosed in braces
then variable substitution will occur.  The characters after
the \fB$\fR, up to the first character that isn't a number, letter, or
underscore, are taken as a variable name and the string value of that
variable is substituted for the name.  Or, if the dollar sign is followed
by an open curly brace then the variable name consists of all the characters
up to the next close curly brace.  For example, if variable \fBfoo\fR
has the value \fBtest\fR, then the command
.DS C
\fBvar a $foo.c\fR
.DE
is equivalent to the command
.DS C
\fBvar a test.c\fR
.DE
and the command
.DS C
\fBvar a abc${foo}bar\fR
.DE
is equivalent to the command
.DS C
\fBvar a abctestbar\fR
.DE
Variable substitution does not occur in arguments that are enclosed
in braces:  the
dollar sign and variable name are passed through to the argument verbatim.
.PP
The dollar sign abbreviation is simply a shorthand form.  \fB$a\fR is
completely equivalent to \fB[var a]\fR;  it is provided as a convenience
to reduce typing.

.SH "BACKSLASH SUBSTITUTION"
.PP
Backslashes may be used to insert non-printing characters into
command fields and also to insert braces, brackets, and
dollar signs into fields
without them being interpreted specially as described above.
The backslash sequences understood by the Tcl interpreter are
listed below.  In each case, the backslash
sequence is replaced by the given character:
.TP 20
\fB\eb\fR
Backspace (octal 10).
.TP 20
\fB\ee\fR
Escape (octal 33).
.TP 20
\fB\en\fR
Newline (octal 15).
.TP 20
\fB\et\fR
Tab (octal 11).
.TP 20
\fB\e{\fR
Left brace (``{'').
.TP 20
\fB\e}\fR
Right brace (``}'').
.TP 20
\fB\e[\fR
Open bracket (``['').
.TP 20
\fB\e]\fR
Close bracket (``]'').
.TP 20
\fB\e<space>\fR
Space (`` ''): doesn't terminate argument.
.TP 20
\fB\e\e\fR
Backslash (``\e'').
.TP 20
\fB\eC\fIx\fR
Control-\fIx\fR (\fIx\fR AND octal 037), for any ASCII \fIx\fR except \fBM\fR
(see below).
.TP 20
\fB\eM\fIx\fR
Meta-\fIx\fR (\fIx\fR OR octal 200), for any ASCII \fIx\fR.
.TP 20
\fB\eCM\fIx\fR
Control-meta-\fIx\fR ((\fIx\fR AND octal 037) OR octal 0200), for
any ASCII \fIx\fR.
.TP 20
\fB\e\fIddd\fR
The digits \fIddd\fR (one, two, or three of them) give the octal value of
the character.
.PP
For example, in the command
.DS
\fBvar a \e{x\e[\e\0yz\e141\fR
.DE
the second argument to \fBvar\fR will be ``\fB{x[\0yza\fR''.
.PP
If a backslash is followed by something other than one of the options
described above, then the backslash is transmitted to the argument
field without any special processing, and the Tcl scanner continues
normal processing with the next character.  For example, in the
command
.DS
\fBvar \e*a \e\e\e{foo\fR
.DE
The first argument to \fBvar\fR will be \fB\e*a\fR and the second
argument will be \fB\e{foo\fR.
.PP
If an argument is enclosed in braces, then backslash sequences inside
the argument are parsed but no substitution occurs:  the backslash
sequence is passed through to the argument as is, without making
any special interpretation of the characters in the backslash sequence.
In particular, backslashed braces are not counted in locating the
matching right brace that terminates the argument.
For example, in the
command
.DS
\fBvar a {\e{abc}\fR
.DE
the second argument to \fBvar\fR will be \fB\e{abc\fR.
.PP
This backslash mechanism is not sufficient to generate absolutely
any argument structure;  it only covers the
most common cases.  To produce particularly complicated arguments
it will probably be easiest to use the \fBformat\fR command along with
command substitution.

.SH "COMMAND SUMMARY"
.IP [1]
A command is just a string.
.IP [2]
Within a string commands are normally separated by newlines.
.IP [3]
If a command begins with a bracket then it is terminated by the matching
close bracket and newlines within the command are treated as white space.
.IP [4]
A command consists of fields.  The first field is the name of the command
and other fields are strings that are passed to that command as arguments.
.IP [5]
Fields are normally separated by white space.
.IP [6]
Braces defer interpretation of special characters.
If a field begins with a left brace, then it consists of everything
between the left brace and the matching right brace. The
braces themselves are not included in the argument.
No further processing is done on the information between the braces.
.IP [7]
If a field doesn't begin with a left brace, then backslash,
variable, and command substitution are done on the field.  Only a
single level of processing is done:  the results of one substitution
are not scanned again for further substitutions or any other
special treatment.  Substitution can
occur on \fIany\fR field of a command, including the command name
as well as the arguments.
.IP [8]
If the first non-blank character of a command is a \fB#\fR, everything
from the \fB#\fR up through the next newline is treated as a comment
and ignored.

.SH "EXPRESSIONS"
.PP
The second major interpretation applied to strings in Tcl is
as expressions.  Several commands, such as \fBexpr\fR, \fBfor\fR,
and \fBif\fR, treat some of their arguments as expressions and
call the Tcl expression processor (\fBTcl_Expr\fR) to evaluate them.
A Tcl expression has C-like syntax and evaluates to an integer
result.  Expressions
may contain integer values, variable names in \fB$\fR notation
(the variables' values must be integer strings),
commands (embedded in brackets) that produce integer string results,
parentheses for grouping, and operators.  Numeric values, whether they
are passed directly or through variable or command substitution, may
be specified either in decimal (the normal case), in octal (if the
first character of the value is \fB0\fR), or in hexadecimal (if the first
two characters of the value are \fB0x\fR).
The valid operators are listed
below, grouped in decreasing order of precedence:
.TP 20
\fB\-\0\0~\0\0!\fR
Unary minus, bit-wise NOT, logical NOT.
.TP 20
\fB*\0\0/\0\0%\fR
Multiply, divide, remainder.
.TP 20
\fB+\0\0\-\fR
Add and subtract.
.TP 20
\fB<<\0\0>>\fR
Left and right shift.
.TP 20
\fB<\0\0>\0\0<=\0\0>=\fR
Boolean less, greater, less than or equal, and greater than or equal.
Each operator produces 1 if the condition is true, 0 otherwise.
.TP 20
\fB==\0\0!=\fR
Boolean equal and not equal.  Each operator produces a zero/one result.
.TP 20
\fB&\fR
Bit-wise AND.
.TP 20
\fB^\fR
Bit-wise exclusive OR.
.TP 20
\fB|\fR
Bit-wise OR.
.TP 20
\fB&&\fR
Logical AND.  Produces a 1 result if both operands are non-zero, 0 otherwise.
.TP 20
\fB||\fR
Logical OR.  Produces a 0 result if both operands are zero, 1 otherwise.
.PP
See the C manual for more details on the results
produced by each operator.
All of the binary operators group left-to-right within the same
precedence level.  For example, the expression
.DS
\fB(4*2) < 7\fR
.DE
evaluates to 0.  Evaluating the expression string
.DS
\fB($a + 3) < [var b]\fR
.DE
will cause the values of the variables \fBa\fR and \fBb\fR to be
examined;  the result will be 1
if \fBb\fR is greater than a by at least 3;  otherwise the result
will be 0.
.PP
In general it is safest to enclose an expression in braces when
entering it in a command:  otherwise, if the expression contains
any white space then the Tcl interpreter will split it
among several arguments.  For example, the command
.DS C
\fBexpr $a + $b\fR
.DE
results in three arguments being passed to \fBexpr\fR:  \fB$a\fR,
\fB+\fR, and \fB$b\fR.  In addition, if the expression isn't in braces
then the Tcl interpreter will perform variable and command substitution
immediately (it will happen in the command parser rather than in
the expression parser).  In many cases the expression is being
passed to a command that will evaluate the expression later (or
even many times if, for example, the expression is to be used to
decide when to exit a loop).  Usually the desired goal is to re-do
the variable or command substitutions each time the expression is
evaluated, rather than once and for all at the beginning.  For example,
the command
.DS C
\fBfor {var i 1} $i<=10 {var i [expr $i+1]} {...}\fR
.DE
is probably intended to iterate over all values of \fBi\fR from 1 to 10.

Tools/swat/tcl/tcl.c
/*
 * tcl.c --
 *
 *	Test driver for TCL.
 *
 * Copyright 1987 Regents of the University of California
 * All rights reserved.
 * Permission to use, copy, modify, and distribute this
 * software and its documentation for any purpose and without
 * fee is hereby granted, provided that the above copyright
 * notice appear in all copies.  The University of California
 * makes no representations about the suitability of this
 * software for any purpose.  It is provided "as is" without
 * express or implied warranty.
 */

#ifndef lint
static char rcsid[] = "$Id: tcl.c,v 1.9 97/04/18 12:14:33 dbaumann Exp $ SPRITE (Berkeley)";
#endif not lint

#include <config.h>
#include <compat/string.h>
#include <compat/stdlib.h>

#if defined(_MSDOS) || defined(_WIN32) || defined(_LINUX)
# include <stdio.h>
# include <time.h>
#else  /* not (_MSDOS or _WIN32) */
# include <sys/time.h>
#endif

#include <ctype.h>
#include "setjmp.h"
#include <malloc.h>
#include "tcl.h"

Tcl_Interp *interp;
/*
 * Data for a procedure being debugged
 */
typedef struct {
    int	    	  (*cmdProc)();
    ClientData	  clientData;
    void    	  (*deleteProc)();
} DebugData;

typedef struct _Frame {
    int	    	  (*cmdProc)();
    int	    	  level;
    int	    	  stopOnReturn;
    ClientData	  clientData;
    int	    	  argc;
    char    	  **argv;
    struct _Frame *next;
} Frame;

Frame	    *stack = (Frame *)NULL;
int	    stepping = 0;
int	    skipCalls = 0;
int	    numDebug = 0;
Tcl_Trace   debugTrace;

jmp_buf	    abortBuf;

int
cmdUndebug(ClientData	  	clientData,
	   Tcl_Interp	  	*interp,
	   int	    	  	argc,
	   char    	  	**argv);

void
debugClean(DebugData *dd)
{
    free((char *)dd);

    numDebug--;

    if (numDebug == 0) {
	Tcl_DeleteTrace(interp, debugTrace);
	debugTrace = NULL;
    }
}

int
debugCatch(DebugData *dd,
	   Tcl_Interp interp,
	   int argc,
	   char **argv)
{
    return ((* dd->cmdProc) (dd->clientData, interp, argc, argv));
}

void
debugFrame(Frame *f, int abbrev)
{
    int	    	  i;

    printf("%c [%s", f->stopOnReturn ? 'b' : ' ', f->argv[0]);

    for (i = 1; i < f->argc; i++) {
	int 	  j;
	char 	  *cp;

	for (cp = f->argv[i]; *cp; cp++) {
	    if (isspace(*cp)) {
		break;
	    }
	}

	if (*cp || cp == f->argv[i]) {
	    printf(" {");
	    for (j = 0, cp = f->argv[i]; (!abbrev || j < 10) && *cp; cp++,j++){
		if (*cp == '\n') {
		    printf("\\n");
		    j -= 2;
		} else if (*cp == '\t') {
		    printf("\\t");
		    j -= 2;
		} else {
		    putchar(*cp);
		}
	    }
	    if (*cp) {
		printf("...}");
	    } else {
		printf("}");
	    }
	} else {
	    printf(" %s", f->argv[i]);
	}
    }
    printf("]\n");
}

void
doDebug(void)
{
    while(1) {
	char	  cmd[1000], *p;
	int 	  i, result;

	clearerr(stdin);

	printf("[tcl] ");
	fflush(stdout);

	gets(cmd);
	p = index(cmd, ' ');
	if (p == (char *)NULL) {
	    p = cmd + strlen(cmd);
	}
	i = p - cmd;

	if (strncmp(cmd, "step", i) == 0) {
	    stepping = 1;
	    skipCalls = 0;
	    break;
	} else if (strncmp(cmd, "cont", i) == 0) {
	    stepping = 0;
	    skipCalls = 0;
	    break;
	} else if (strncmp(cmd, "eval", i) == 0) {
	    char	*oldResult;
	    int	result;

	    stepping = skipCalls = 0;
	    oldResult = malloc(strlen(interp->result) + 1);
	    strcpy(oldResult, interp->result);

	    result = Tcl_Eval(interp, p, 0, 0);
	    if (result == TCL_OK) {
		if (*interp->result != 0) {
		    printf("%s\n", interp->result);
		}
	    } else {
		if (result == TCL_ERROR) {
		    printf("Error");
		} else {
		    printf("Returned code %d", result);
		}
		if (*interp->result != 0) {
		    printf(": %s\n", interp->result);
		} else {
		    printf("\n");
		}
	    }
	    Tcl_Return(interp, oldResult, TCL_DYNAMIC);
	} else if (strncmp(cmd, "next", i) == 0) {
	    stepping = 0; skipCalls = 1;
	    break;
	} else if (strncmp(cmd, "where", i) == 0) {
	    Frame	*f;

	    for (f = stack; f != (Frame *)NULL; f = f->next) {
		debugFrame(f, 1);
	    }
	} else if (strncmp(cmd, "frame", i) == 0) {
	    debugFrame(stack, 0);
	} else if (strncmp(cmd, "quit", i) == 0) {
	    Frame *f;

	    skipCalls = 1;
	    stepping = 0;
	    for (f = stack; f != (Frame *)NULL; f = f->next) {
		f->stopOnReturn = 0;
	    }
	    break;
	} else if (strncmp(cmd, "abort", i) == 0) {
	    Frame *f;

	    while(stack) {
		f = stack->next;
		free((char *)stack);
		stack = f;
	    }
	    stepping = 0;
	    skipCalls = 0;
	    Tcl_TopLevel(interp);
	    longjmp(abortBuf, 1);
	} else if (strncmp(cmd, "undebug", i) == 0) {
	    char  *argv[3];

	    argv[0] = "undebug";
	    argv[1] = stack->argv[0];
	    argv[2] = NULL;

	    cmdUndebug(NULL, interp, 2, argv);
	} else if (strncmp(cmd, "help", i) == 0) {
	    printf("Available commands:\n");
	    printf("\ts[tep]    - Step to next tcl call\n");
	    printf("\tn[ext]    - Finish current call, ignoring nested calls\n");
	    printf("\t            and breakpoints\n");
	    printf("\tc[ont]    - Finish current call, but stop at any breakpoints\n");
	    printf("\tw[here]   - Print current call stack\n");
	    printf("\th[elp]    - Get this message\n");
	    printf("\tf[rame]   - Print current frame in long form\n");
	    printf("\te[val]    - Evaluate command in interpreter. Stops at any\n");
	    printf("\t            breakpoints you've set.\n");
	    printf("\tq[uit]    - Get back to top level, finishing out all\n");
	    printf("\t            current frames without stopping.\n");
	    printf("\ta[bort]   - Get back to top level without finishing\n");
	    printf("\t            current calls\n");
	    printf("\tu[ndebug] - Remove a breakpoint from the current frame\n");
	} else {
	    printf("Unknown command: %.*s\n", i, cmd);
	}
    }
}

void
debugCallProc(ClientData  cd,
	      Tcl_Interp  *interp,
	      int	  level,
	      char	  *command,
	      int	  (*cmdProc)(),
	      ClientData  cmdData,
	      int	  argc,
	      char	  **argv)
{
    Frame	  *top;

    top = (Frame *)malloc(sizeof(Frame));
    top->cmdProc = cmdProc;
    top->clientData = cmdData;
    top->argc = argc;
    top->argv = argv;
    top->next = stack;
    top->level = level;
    top->stopOnReturn = 0;
    stack = top;

    if (((cmdProc == debugCatch) && !skipCalls) || stepping) {
	/*
	 * Either we hit a function being debugged and weren't told
	 * to skip over calls, or we're single stepping, so we decide
	 * to stop and wait for further orders.
	 */
	printf("stopped in %s\n", argv[0]);

	stack->stopOnReturn = 1;
	debugFrame(stack, 1);

	doDebug();
    }
}

void
debugReturnProc(ClientData	cd,
		Tcl_Interp	*interp,
		int		level,
		char		*command,
		int		(*cmdProc)(),
		ClientData	cmdData,
		int		result)
{
    Frame	  *top;

    /*
     * We'll get called on the return from cmdDebug, but we won't have a
     * stack set up at that point, so we still need to check to make sure
     * we actually have a stack at this point.
     */
    if (stack) {
	if (stack->stopOnReturn) {
	    printf("%s returning ", stack->argv[0]);
	    switch(result) {
	    case TCL_OK:
		printf("\"%s\"\n", interp->result);
		break;
	    case TCL_ERROR:
		printf("error \"%s\"\n", interp->result);
		break;
	    case TCL_BREAK:
		printf("break\n");
		break;
	    case TCL_CONTINUE:
		printf("continue\n");
		break;
	    default:
		printf("code %d \"%s\"\n", result, interp->result);
		break;
	    }

	    doDebug();
	}

	top = stack->next;
	free((char *)stack);
	stack = top;
	if (stack == (Frame *)NULL) {
	    skipCalls = 0;
	}
    }
}

int
cmdDebug(ClientData	  	clientData,
	 Tcl_Interp	  	*interp,
	 int	    	  	argc,
	 char    	  	**argv)
{
#if 0
    int	    	  	i;
    DebugData	  	*dd;

    for (i = 1; i < argc; i++) {
	dd = (DebugData *)malloc(sizeof(DebugData));
	const char *realName;

	if (!Tcl_FetchCommand(interp, argv[i], &realName,
			      &dd->cmdProc, &dd->clientData,
			      &dd->deleteProc))
	{
	    Tcl_RetPrintf(interp, "%s: not defined", argv[i]);
	    return(TCL_ERROR);
	}

	if (dd->cmdProc == debugCatch) {
	    printf("%s: already being debugged\n", argv[i]);
	} else {
	    Tcl_OverrideCommand(interp, realName,
				debugCatch, (ClientData)dd, debugClean,
				&dd->cmdProc, &dd->clientData,
				&dd->deleteProc);
	    numDebug++;
	}
    }

    if (debugTrace == (Tcl_Trace)NULL) {
	debugTrace = Tcl_CreateTrace(interp, 0, debugCallProc, debugReturnProc,
				     (ClientData)NULL);
    }
#endif
    return(TCL_OK);
}

int
cmdUndebug(ClientData	  	clientData,
	   Tcl_Interp	  	*interp,
	   int	    	  	argc,
	   char    	  	**argv)
{
#if 0
    int	    	  	i;

    for (i = 1; i < argc; i++) {
	int 	  	(*cmdProc)();
	ClientData	clientData;
	void		(*deleteProc)();
	DebugData 	*dd;
	const char  	*realName;

	if (!Tcl_FetchCommand(interp, argv[i], &realName,
			      &cmdProc, &clientData,
			      &deleteProc))
	{
	    Tcl_RetPrintf(interp, "%s: not defined", argv[i]);
	    return(TCL_ERROR);
	}

	if (cmdProc != debugCatch) {
	    printf("%s: not being debugged\n", argv[i]);
	} else {
	    dd = (DebugData *)clientData;

	    Tcl_OverrideCommand(interp, realName,
				dd->cmdProc, dd->clientData, dd->deleteProc,
				&cmdProc, &clientData, &deleteProc);
	    debugClean(dd);
	}
    }
#endif
    return(TCL_OK);
}

int
cmdEcho(ClientData clientData,
	Tcl_Interp *interp,
	int argc,
	char **argv)
{
    int i;

    for (i = 1; ; i++) {
	if (argv[i] == NULL) {
	    if (i != argc) {
	    echoError:
		Tcl_RetPrintf(interp,
		    "argument list wasn't properly NULL-terminated in \"%s\" command",
		    argv[0]);
	    }
	    break;
	}
	if (i >= argc) {
	    goto echoError;
	}
	fputs(argv[i], stdout);
	if (i < (argc-1)) {
	    printf(" ");
	}
    }
    printf("\n");
    return TCL_OK;
}

void
deleteProc(ClientData clientData)
{
    printf("Deleting command with clientData \"%s\".\n", clientData);
}

int
cmdDelete(ClientData clientData,
	  Tcl_Interp *interp,
	  int argc,
	  char **argv)
{
    if (argc != 2) {
	interp->result = "syntax: delete command";
	return TCL_ERROR;
    }

    if (strcmp(argv[1], "*") == 0) {
	Tcl_DeleteInterp(interp);
    } else {
	Tcl_DeleteCommand(interp, argv[1]);
    }
    return TCL_OK;
}

void
traceCallProc(ClientData clientData,
	      Tcl_Interp *interp,
              Tcl_Frame *frame)
{
    int i;

    printf("Level %d, clientData 0x%x, interp 0x%x, calling \"%s\"\n",
	   frame->level, clientData, interp, frame->command);
    printf("    cmdProc 0x%x, cmdClientData 0x%x\n",
	   frame->cmdProc, frame->cmdData);
    for (i = 0; i < frame->argc; i++) {
	printf("        argv[%d] = \"%s\"\n", i, frame->argv[i]);
    }

}

void
traceReturnProc(ClientData clientData,
		Tcl_Interp *interp,
                Tcl_Frame *frame,
		int result)
{
    switch (result) {
    case TCL_OK:
	printf ("%s returns \"%s\"\n", frame->command, interp->result);
	break;
    case TCL_ERROR:
	printf("%s returns error \"%s\"\n", frame->command, interp->result);
	break;
    case TCL_BREAK:
	printf("%s returns BREAK\n", frame->command);
	break;
    case TCL_CONTINUE:
	printf("%s returns CONTINUE\n", frame->command);
	break;
    }
}

int
cmdTrace(ClientData clientData,		/* Not used. */
	 Tcl_Interp *interp,
	 int argc,
	 char *argv[])
{
    int length, level;
    Tcl_Trace trace;

    if (argc != 3) {
	interp->result = "wrong # args: should be \"trace create|delete arg\"";
	return TCL_ERROR;
    }
    if (strcmp(argv[1], "create") == 0) {
	if (sscanf(argv[2], "%d", &level) != 1) {
	    interp->result = "bad arg to \"trace create\"";
	    return TCL_ERROR;
	}

	printf("New trace is 0x%x\n",
		Tcl_CreateTrace(interp, level, traceCallProc,
				traceReturnProc, (ClientData) 47));
	return TCL_OK;
    } else if (strcmp(argv[1], "delete") == 0) {
	if (sscanf(argv[2], "%x", &trace) != 1) {
	    interp->result = "bad arg to \"trace delete\"";
	    return TCL_ERROR;
	}
	Tcl_DeleteTrace(interp, trace);
	return TCL_OK;
    }
    Tcl_RetPrintf(interp, "bad option (%s) to \"trace\": must be create or delete", argv[1]);
    return TCL_ERROR;
}

void
main(void)
{
    char cmd[1000], *p;
    register char *p2;
    int c, i, result;
    extern void exit();

    interp = Tcl_CreateInterp();
    Tcl_CreateCommand(interp, "echo", cmdEcho, 0, (ClientData) "echo",
		      deleteProc);
    Tcl_CreateCommand(interp, "delete", cmdDelete, 0, (ClientData) "delete",
		      deleteProc);
    Tcl_CreateCommand(interp, "trace", cmdTrace, 0, (ClientData) "trace",
		      deleteProc);
    Tcl_CreateCommand(interp, "exit", (Tcl_CmdProc *)exit, 0, (ClientData)0,
		      NoDelProc);
    Tcl_CreateCommand(interp, "debug", cmdDebug, 0, (ClientData)0, NoDelProc);
    Tcl_CreateCommand(interp, "undebug", cmdUndebug, 0, (ClientData)0,
		      NoDelProc);

    Tcl_SetVar(interp, "prompt", "% ", 1);

    (void)setjmp(abortBuf);

    while (1) {
	const char  *prompt = Tcl_GetVar(interp, "prompt", 1);

	clearerr(stdin);
	fputs(prompt, stdout);
	fflush(stdout);
	p = cmd;
	while (1) {
	    c = getchar();
	    if (c == EOF) {
		if (p == cmd) {
		    exit(0);
		}
		goto gotCommand;
	    }
	    *p = c;
	    p++;
	    if (c == '\n') {
		register char *p2;
		int parens, brackets, numBytes;

		for (p2 = cmd, parens = 0, brackets = 0; p2 != p; p2++) {
		    switch (*p2) {
			case '\\':
			    Tcl_Backslash(p2, &numBytes);
			    p2 += numBytes-1;
			    break;
			case '{':
			    parens++;
			    break;
			case '}':
			    parens--;
			    break;
			case '[':
			    brackets++;
			    break;
			case ']':
			    brackets--;
			    break;
		    }
		}
		if ((parens <= 0) && (brackets <= 0)) {
		    goto gotCommand;
		}
	    }
	}
	gotCommand:
	*p = 0;

	result = Tcl_Eval(interp, cmd, 0, &p);
	if (result == TCL_OK) {
	    if (*interp->result != 0) {
		printf("%s\n", interp->result);
	    }
	} else if ((result == TCL_ERROR) &&
		   strncmp(interp->result, "invoked",
			   sizeof("invoked") - 1) == 0)
	{
	    char  *newCmd;

	    newCmd = malloc(strlen("exec ") + strlen(cmd) + 1);

	    sprintf(newCmd, "exec %s", cmd);
	    result = Tcl_Eval(interp, newCmd, 0, (char **)NULL);
	    if (result == TCL_ERROR) {
		printf("Error ");
	    }
	    if (interp->result[0] != '\0') {
		printf("%s\n", interp->result);
	    }
	    free(newCmd);
	} else {
	    if (result == TCL_ERROR) {
		printf("Error");
	    } else {
		printf("Returned code %d", result);
	    }
	    if (*interp->result != 0) {
		printf(": %s\n", interp->result);
	    } else {
		printf("\n");
	    }
	}
    }
}
